File: data/monsters/monster.lua
Monster = {}

function Monster:new()
  new_monster = { health = 20, attack_damage = 10 }
  setmetatable(new_monster, self)
  self.__index = self
  return new_monster
end

function Monster:get_stats()
  return { health = self.health }
end

function Monster:description()
  return "Large monster"
end

function Monster:attack()
  return self.attack_damage, "The monster farts in your general direction"
end

function Monster:take_damage(damage)
  if damage > self.health then
    damage = self.health
    self.health = 0
  else
    self.health = self.health - damage
  end
  result = "Monster takes " .. damage .. " points of damage"
  if self.health == 0 then
    result = result .. "\nMonster is dead!"
  end
  return result
end

File: monster/monster.cpp
#include "monster.h"
#include <sstream>

void Monster::init() {
    this->health = 50;
}

Stats Monster::getStats() const {
    Stats stats;
    stats.insert({ std::string("health"), std::to_string(this->health) });
    return stats;
}

StyledText::Text& Monster::getDescription() const {
    return StyledText::text("Large monster");
}

bool Monster::isAlive() const {
    return this->health > 0;
}

AttackResult Monster::attack() {
    return { 10, StyledText::text("The monster", StyledText::green("farts"), "in your general direction") };
}

AttackResult Monster::takeDamage(Damage damage) {
    if (damage > this->health) {
        damage = this->health;
        this->health = 0;
    } else {
        this->health -= damage;
    }
    return { damage, StyledText::text("The monster takes", StyledText::red(std::to_string(damage)), "points of damage") };
}
